src/wallet/rpcwallet.cpp

```diff
@@ -4932,36 +4871,88 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+                                },
+                                [](const auto &) { }
+                        }, preferredRecipient.value());
+                    }
+                },
+                [](const auto &) { }
+        }, addr.value());
 
-        recipients.push_back(ResolvedPayment(ua, addr.value(), nAmount, memo));
+        recipients.push_back(Payment(addr.value(), nAmount, memo));
         nTotalOut += nAmount;
     }
     if (recipients.empty()) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "No recipients");
     }
 
+    // Check that the from address is valid.
+    // Unified address (UA) allowed here (#5185)
+    auto fromaddress = params[0].get_str();
+    ZTXOSelector ztxoSelector = [&]() {
+        if (fromaddress == "ANY_TADDR") {
+            return CWallet::LegacyTransparentZTXOSelector(true, TransparentCoinbasePolicy::Disallow);
+        } else {
+            auto decoded = keyIO.DecodePaymentAddress(fromaddress);
+            if (!decoded.has_value()) {
+                throw JSONRPCError(
+                        RPC_INVALID_ADDRESS_OR_KEY,
+                        "Invalid from address: should be a taddr, zaddr, UA, or the string 'ANY_TADDR'.");
+            }
+
+            auto ztxoSelectorOpt = pwalletMain->ZTXOSelectorForAddress(
+                decoded.value(),
+                true,
+                tcoinbasePolicy,
+                // LegacyCompat does not include AllowLinkingAccountAddresses.
+                maybeStrategy.has_value() ? maybeStrategy.value().AllowLinkingAccountAddresses() : false);
+            if (!ztxoSelectorOpt.has_value()) {
+                throw JSONRPCError(
+                        RPC_INVALID_ADDRESS_OR_KEY,
+                        "Invalid from address, no payment source found for address.");
+            }
+
+            auto selectorAccount = pwalletMain->FindAccountForSelector(ztxoSelectorOpt.value());
+            std::visit(match {
+                [&](const libzcash::UnifiedAddress& ua) {
+                    if (!selectorAccount.has_value() || selectorAccount.value() == ZCASH_LEGACY_ACCOUNT) {
+                        throw JSONRPCError(
+                                RPC_INVALID_ADDRESS_OR_KEY,
+                                "Invalid from address, UA does not correspond to a known account.");
+                    }
+                    involvesUnifiedAddress = true;
+                },
+                [&](const auto& other) {
+                    if (selectorAccount.has_value() && selectorAccount.value() != ZCASH_LEGACY_ACCOUNT) {
+                        throw JSONRPCError(
+                                RPC_INVALID_ADDRESS_OR_KEY,
+                                "Invalid from address: is a bare receiver from a Unified Address in this wallet. Provide the UA as returned by z_getaddressforaccount instead.");
+                    }
+                }
+            }, decoded.value());
```

Comment:
```suggestion
            bool unknown_or_legacy = !selectorAccount.has_value() || selectorAccount.value() == ZCASH_LEGACY_ACCOUNT;
            std::visit(match {
                [&](const libzcash::UnifiedAddress& ua) {
                    if (unknown_or_legacy) {
                        throw JSONRPCError(
                                RPC_INVALID_ADDRESS_OR_KEY,
                                "Invalid from address, UA does not correspond to a known account.");
                    }
                    involvesUnifiedAddress = true;
                },
                [&](const auto& other) {
                    if (!unknown_or_legacy) {
                        throw JSONRPCError(
                                RPC_INVALID_ADDRESS_OR_KEY,
                                "Invalid from address: is a bare receiver from a Unified Address in this wallet. Provide the UA as returned by z_getaddressforaccount instead.");
                    }
                }
            }, decoded.value());
```

Reply:
Hrmm, does GH not show diff refinement on suggestions? I guess I don’t really notice on smaller ones, but this was a pain. Was obvious once I pasted it into my editor.